### DISCUSSION TITLE
Rename Tabs

### DISCUSSION THREAD

**Discussion by efendioglukdz** [Adanos020/egui_dock#135] [Q&A] [answered]
Is there any possibility to rename an active Tab?

**Comment by lvaroqui** [comment]
I'm not sure what you mean, but you provide the tab name through the [title](https://docs.rs/egui_dock/latest/egui_dock/widgets/tab_viewer/trait.TabViewer.html#tymethod.title) API.

And you can find the active tab with the [find_active_focus](https://docs.rs/egui_dock/latest/egui_dock/tree/struct.Tree.html#method.find_active_focused) API.

**Reply by efendioglukdz** [reply]
Imagine I am writing a texteditor with files to open. Each opened new file is open on seperate Tab. So the tab would be called unnamed.txt initially but after save as it gets another name, so the name of the tab must change too

**Reply by lvaroqui** [reply]
That's what the title API is for, you can provide any title you like for the tab with it!

**Reply by lvaroqui** [reply, marked as answer]
Something like this:
```rust
struct FileTab {
    title: String,
    content: String,
}

impl egui_dock::TabViewer for TabViewer {
    type Tab = FileTab;

    fn ui(&mut self, ui: &mut egui::Ui, tab: &mut Self::Tab) {
        egui::TextEdit::multiline(&mut self.content);
    }

    fn title(&mut self, tab: &mut Self::Tab) -> egui::WidgetText {
        self.title.into()
    }
}
```

**Reply by anhosh** [reply]
I'd just like to add that by default `TabViewer::id` returns an ID derived from `TabViewer::title`. This means that if your tab's title changes in runtime, its ID will also change, which can lead to weird issues like [your tab losing focus](https://github.com/Adanos020/egui_dock/issues/158).

So, if you want your tab titles to change, you have to override `TabViewer::id` and derive the returned ID on something that doesn't change :)
